datapreprocess/ray/tokenize_shuffle.py

```diff
@@ -1,78 +1,231 @@
+SAMPLING_FREQUENCIES[Sources.ARXIV] = 1.080357143
+SAMPLING_FREQUENCIES[Sources.STACKEXCHANGE] = 1.21
+SAMPLING_FREQUENCIES[Sources.UNKNOWN] = 0
 
-    # Use the head_object call to get the total size of the object
-    meta_data = s3.head_object(Bucket=bucket_name, Key=key)
-    total_size = int(meta_data.get("ContentLength", 0))
 
-    # Create a progress bar with tqdm
-    progress = tqdm(total=total_size, unit="B", unit_scale=True)
+def jsonl_file_reader(fh: BinaryIO, content_key: str):
+    with io.TextIOWrapper(fh, encoding="utf-8") as text_reader:
+        with jsonlines.Reader(text_reader) as jsonl_reader:
+            for item in jsonl_reader:
+                yield item[content_key]
 
-    # Callback to update the progress bar
-    def progress_callback(bytes_transferred):
-        progress.update(bytes_transferred)
 
-    # Use BytesIO to store the downloaded bytes in memory
-    buffer = BytesIO()
-    s3.download_fileobj(bucket_name, key, buffer, Callback=progress_callback)
+def zstd_compressed_reader(fh: BinaryIO, content_key: str):
+    dctx = zstd.ZstdDecompressor()
+    with dctx.stream_reader(fh) as reader:
+        for item in jsonl_file_reader(reader, content_key=content_key):
+            yield item
 
-    progress.close()
 
-    # Reset the buffer's position to the beginning
-    buffer.seek(0)
+def gzip_compressed_reader(fh: BinaryIO, content_key: str):
+    with gzip.open(fh, "rb") as f_in:
+        with jsonlines.Reader(f_in) as jsonl_reader:
+            for item in jsonl_reader:
+                yield item[content_key]
 
-    # Return the in-memory buffer
-    return buffer.read()
+
+def get_reader(file_type, content_key: str):
+    if file_type == RawFileType.JSONL:
+        return lambda x: jsonl_file_reader(x, content_key=content_key)
+    if file_type == RawFileType.ZSTD_JSONL_COMPRESSED:
+        return lambda x: zstd_compressed_reader(x, content_key=content_key)
+    if file_type == RawFileType.GZIP_JSONL_COMPRESSED:
+        return lambda x: gzip_compressed_reader(x, content_key=content_key)
+    else:
+        raise Exception("Unsupported filetype")
+
+
+def get_raw_filetype(key: str):
+    if key.endswith(".jsonl") or key.endswith(".json"):
+        return RawFileType.JSONL
+    elif key.endswith(".jsonl.zst") or key.endswith(".json.zst"):
+        return RawFileType.ZSTD_JSONL_COMPRESSED
+    elif key.endswith(".jsonl.gz") or key.endswith(".json.gz"):
+        return RawFileType.GZIP_JSONL_COMPRESSED
+    else:
+        logger.warning(f"Unknown filetype: {key}")
+        return RawFileType.UNKNOWN
+
+
+def preprocess(
+    key: str,
+    fh: BinaryIO,
+    seed: int,
+    content_key: str,
+    seqlen: int = 8192,
+    tokenizer=None,
+    do_sample: bool = False,
+):
+    tokenizer_fn, vocab_size = tokenizer
+    rng = random.Random(hash(key) + seed)
+    EOT = SpecialTokens.END_OF_TEXT.value % (vocab_size + len(SpecialTokens))
+    PAD = SpecialTokens.PAD.value % (vocab_size + len(SpecialTokens))
```

Comment:
This implicitly extends the vocabulary size, right? Seems like it should work, but I'm not sure if it causes issues down the line.

Reply:
yes this is the behavior we want.

Reply:
Ok thanks!